packages/compiler-cli/src/ngtsc/annotations/src/component.ts

```diff
@@ -1431,3 +1443,35 @@ function makeCyclicImportInfo(
       `The ${type} '${name}' is used in the template but importing it would create a cycle: `;
   return makeRelatedInformation(ref.node, message + path);
 }
+
+
+/**
+ * Checks whether a selector is a valid custom element tag name.
+ * Based loosely on https://github.com/sindresorhus/validate-element-name.
+ */
+function checkCustomElementSelectorForErrors(selector: string): string|null {
+  // Avoid flagging components with an attribute or class selector. This isn't bulletproof since it
+  // won't catch cases like `foo[]bar`, but we don't need it to be. This is mainly to avoid flagging
+  // something like `foo-bar[baz]` incorrectly.
+  if (selector.includes('.') || (selector.includes('[') && selector.includes(']'))) {
+    return null;
+  }
+
+  if (!selector.includes('-')) {
+    return 'Selector of a component that uses ViewEncapsulation.ShadowDom must contain a hyphen.';
+  }
+
+  if (selector.startsWith('-')) {
+    return 'Selector of a ShadowDom-encapsulated component must not start with a hyphen.';
+  }
+
+  if (/[A-Z]/.test(selector)) {
+    return 'Selector of a ShadowDom-encapsulated component must be in lower case.';
+  }
+
+  if (/^\d/i.test(selector)) {
+    return 'Selector of a ShadowDom-encapsulated component must not start with a digit.';
+  }
```

Comment:
I feel like the error message should be guiding people to do the correct thing. So rather than "it should not have xxx" we should be saying "it should have YYY". For example:

```suggestion
  if (/^[^a-z]/.test(selector)) {
    return 'Selector of a ShadowDom-encapsulated component must start with a lower case letter.';
  }

  if (/[A-Z]/.test(selector)) {
    return 'Selector of a ShadowDom-encapsulated component must all be in lower case.';
  }
```